core/chains/evm/logpoller/orm.go

```diff
@@ -300,360 +344,403 @@ func (o *DbORM) SelectLogsCreatedAfter(address common.Address, eventSig common.H
+	args, err := newQueryArgs(o.chainID).
+		withAddress(address).
+		withTopicIndex(topicIndex).
+		withStartBlock(startBlock).
+		withEndBlock(endBlock).
+		withCustomHashArg("sigA", sigA).
+		withCustomHashArg("sigB", sigB).
+		withConfs(confs).
+		toArgs()
+	if err != nil {
 		return nil, err
 	}
 
-	q := o.q.WithOpts(qopts...)
-	var logs []Log
-
-	err := q.Select(&logs, `
-		SELECT *
-		FROM   evm.logs
-		WHERE  evm_chain_id = $1
-		AND    address = $2
-		AND    event_sig = $3
-		AND block_number BETWEEN $6 AND $7
-		AND block_number <= (SELECT COALESCE(block_number, 0) FROM evm.log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1) - $8
-		
+	nestedQuery := nestedBlockNumberQuery()
+	query := fmt.Sprintf(`
+		SELECT * FROM   evm.logs
+		WHERE   evm_chain_id = :evm_chain_id
+		AND     address = :address
+		AND     event_sig = :sigA
+		AND 	block_number BETWEEN :start_block AND :end_block
+		AND 	block_number <= %s		
 		EXCEPT
-		
-		SELECT     a.*
-		FROM       evm.logs AS a
+		SELECT     a.* FROM       evm.logs AS a
 		INNER JOIN evm.logs B
 		ON         a.evm_chain_id = b.evm_chain_id
 		AND        a.address = b.address
-		AND        a.topics[$5] = b.topics[$5]
-		AND        a.event_sig = $3
-		AND        b.event_sig = $4
-	    AND 	   b.block_number BETWEEN $6 AND $7
-		AND		   b.block_number <= (SELECT COALESCE(block_number, 0) FROM evm.log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1) - $8
-
-		ORDER BY block_number,log_index ASC
-			`, utils.NewBig(o.chainID), address, sigA.Bytes(), sigB.Bytes(), topicIndex+1, startBlock, endBlock, confs)
-	if err != nil {
+		AND        a.topics[:topic_index] = b.topics[:topic_index]
+		AND        a.event_sig = :sigA
+		AND        b.event_sig = :sigB
+	    AND 	   b.block_number BETWEEN :start_block AND :end_block
+		AND		   b.block_number <= %s
+		ORDER BY block_number,log_index ASC`, nestedQuery, nestedQuery)
+	var logs []Log
+	if err := o.q.WithOpts(qopts...).SelectNamed(&logs, query, args); err != nil {
 		return nil, err
 	}
 	return logs, nil
 }
 
-func (o *DbORM) blocksRangeAfterTimestamp(after time.Time, confs int, qopts ...pg.QOpt) (int64, int64, error) {
-	type blockRange struct {
-		MinBlockNumber int64 `db:"min_block"`
-		MaxBlockNumber int64 `db:"max_block"`
+func (o *DbORM) blocksRangeAfterTimestamp(after time.Time, confs Confirmations, qopts ...pg.QOpt) (int64, int64, error) {
+	args, err := newQueryArgs(o.chainID).
+		withBlockTimestampAfter(after).
+		toArgs()
+	if err != nil {
+		return 0, 0, err
 	}
 
-	var br blockRange
-	q := o.q.WithOpts(qopts...)
-	err := q.Get(&br, `
-		SELECT 
-		    coalesce(min(block_number), 0) as min_block, 
-		    coalesce(max(block_number), 0) as max_block
-		FROM evm.log_poller_blocks 
-		WHERE evm_chain_id = $1
-		AND block_timestamp > $2`, utils.NewBig(o.chainID), after)
```

Comment:
I'm not sure I understand the refactor of this query.  Why do we request a whole `LogPollerBlock` structure if we only need the block number?  The new query looks more complex, are we sure it's still just as fast?

Reply:
You are right, I was "too fast" and these changes are required for finality, not needed to be there. Bear with me

Reply:
It does not introduce any [performance degradation](https://github.com/smartcontractkit/chainlink/pull/10743)
```
BenchmarkSelectLogsCreatedAfter
BenchmarkSelectLogsCreatedAfter-12    	      78	  14378400 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      81	  14428778 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      75	  14412032 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      78	  14276582 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      75	  14363945 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      80	  14359052 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      82	  14357271 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      81	  14418366 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      78	  14398630 ns/op
BenchmarkSelectLogsCreatedAfter-12    	      78	  14530038 ns/op
```

Reply:
Query plan might look a bit complicated to the previous one, but it's only because of the nest query and sort. We apply these on two elements, and then pick LogPollerBlocks by the primary_key which is extremely fast


```sql
 chainlink_test=# explain analyze SELECT * FROM evm.log_poller_blocks
                WHERE evm_chain_id = 1
                AND block_number in (
                    SELECT unnest(array[min(block_number), max(block_number)]) FROM evm.log_poller_blocks
                                WHERE evm_chain_id = 1
                                AND block_timestamp > '1990-10-10')
                order by block_number
chainlink_test-# ;
                                                                                                   QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=849.55..849.64 rows=34 width=62) (actual time=0.672..0.672 rows=2 loops=1)
   Sort Key: log_poller_blocks.block_number
   Sort Method: quicksort  Memory: 25kB
   ->  Nested Loop  (cost=3.36..848.69 rows=34 width=62) (actual time=0.511..0.583 rows=2 loops=1)
         ->  HashAggregate  (cost=2.94..3.94 rows=100 width=8) (actual time=0.402..0.403 rows=2 loops=1)
               Group Key: unnest(ARRAY[($0), ($1)])
               ->  ProjectSet  (cost=1.17..1.69 rows=100 width=8) (actual time=0.393..0.394 rows=2 loops=1)
                     InitPlan 1 (returns $0)
                       ->  Limit  (cost=0.43..0.58 rows=1 width=8) (actual time=0.240..0.241 rows=1 loops=1)
                             ->  Index Scan using log_poller_blocks_pkey on log_poller_blocks log_poller_blocks_1  (cost=0.43..103903.70 rows=662098 width=8) (actual time=0.237..0.237 rows=1 loops=1)
                                   Index Cond: ((block_number IS NOT NULL) AND (evm_chain_id = '1'::numeric))
                                   Filter: (block_timestamp > '1990-10-10 00:00:00+00'::timestamp with time zone)
                     InitPlan 2 (returns $1)
                       ->  Limit  (cost=0.43..0.58 rows=1 width=8) (actual time=0.120..0.121 rows=1 loops=1)
                             ->  Index Scan Backward using log_poller_blocks_pkey on log_poller_blocks log_poller_blocks_2  (cost=0.43..103903.70 rows=662098 width=8) (actual time=0.120..0.120 rows=1 loops=1)
                                   Index Cond: ((block_number IS NOT NULL) AND (evm_chain_id = '1'::numeric))
                                   Filter: (block_timestamp > '1990-10-10 00:00:00+00'::timestamp with time zone)
                     ->  Result  (cost=1.17..1.18 rows=1 width=16) (actual time=0.369..0.369 rows=1 loops=1)
         ->  Index Scan using idx_evm_log_poller_blocks_order_by_block on log_poller_blocks  (cost=0.43..8.45 rows=1 width=62) (actual time=0.087..0.087 rows=1 loops=2)
               Index Cond: ((evm_chain_id = '1'::numeric) AND (block_number = (unnest(ARRAY[($0), ($1)]))))
 Planning Time: 1.496 ms
 Execution Time: 0.896 ms
(22 rows)

chainlink_test=# explain analyze SELECT * FROM evm.log_poller_blocks
                WHERE evm_chain_id = 1
                AND block_number in (
                    SELECT unnest(array[min(block_number), max(block_number)]) FROM evm.log_poller_blocks
                                WHERE evm_chain_id = 1
                                AND block_timestamp > '2090-10-10')
                order by block_number
;
                                                                                              QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=1110.14..1110.22 rows=34 width=62) (actual time=0.328..0.328 rows=0 loops=1)
   Sort Key: log_poller_blocks.block_number
   Sort Method: quicksort  Memory: 25kB
   ->  Nested Loop  (cost=263.95..1109.27 rows=34 width=62) (actual time=0.312..0.312 rows=0 loops=1)
         ->  HashAggregate  (cost=263.52..264.52 rows=100 width=8) (actual time=0.304..0.305 rows=1 loops=1)
               Group Key: unnest(ARRAY[(min(log_poller_blocks_1.block_number)), (max(log_poller_blocks_1.block_number))])
               ->  ProjectSet  (cost=261.75..262.27 rows=100 width=8) (actual time=0.291..0.291 rows=2 loops=1)
                     ->  Aggregate  (cost=261.75..261.76 rows=1 width=16) (actual time=0.280..0.280 rows=1 loops=1)
                           ->  Index Scan using log_poller_blocks_by_timestamp on log_poller_blocks log_poller_blocks_1  (cost=0.43..261.41 rows=68 width=8) (actual time=0.273..0.273 rows=0 loops=1)
                                 Index Cond: ((evm_chain_id = '1'::numeric) AND (block_timestamp > '2090-10-10 00:00:00+00'::timestamp with time zone))
         ->  Index Scan using idx_evm_log_poller_blocks_order_by_block on log_poller_blocks  (cost=0.43..8.45 rows=1 width=62) (actual time=0.003..0.003 rows=0 loops=1)
               Index Cond: ((evm_chain_id = '1'::numeric) AND (block_number = (unnest(ARRAY[(min(log_poller_blocks_1.block_number)), (max(log_poller_blocks_1.block_number))]))))
 Planning Time: 0.736 ms
 Execution Time: 0.477 ms
(14 rows)
```